libraries/AP_ADSB/AP_ADSB_Sagetech.cpp

```diff
@@ -0,0 +1,746 @@
+
+    } else {
+        // use integrated altitude - recommend by sagetech
+        pkt.payload[2] = 0x80;
+        pkt.payload[3] = 0x00;
+    }
+
+
+    // RF mode
+    pkt.payload[4] = last_operating_rf_select;
+
+    send_msg(pkt);
+}
+
+void AP_ADSB_Sagetech::send_GPS()
+{
+    Packet_XP pkt {};
+
+    pkt.type = MsgTypes_XP::GPS_Set;
+    pkt.payload_length = 52;
+    pkt.id = 0;
+
+    const int32_t longitude = frontend._my_loc.lng;
+    const int32_t latitude =  frontend._my_loc.lat;
+
+    // longitude and latitude
+    // NOTE: these MUST be done in double or else we get roundoff in the maths
+    const double lon_deg = longitude * (double)1.0e-7 * (longitude < 0 ? -1 : 1);
+    const double lon_minutes = (lon_deg - int(lon_deg)) * 60;
+    snprintf((char*)&pkt.payload[0], 12, "%03u%02u.%05u", (unsigned)lon_deg, (unsigned)lon_minutes, unsigned((lon_minutes - (int)lon_minutes) * 1.0E5));
+
+    const double lat_deg = latitude * (double)1.0e-7 * (latitude < 0 ? -1 : 1);
+    const double lat_minutes = (lat_deg - int(lat_deg)) * 60;
+    snprintf((char*)&pkt.payload[11], 11, "%02u%02u.%05u", (unsigned)lat_deg, (unsigned)lat_minutes, unsigned((lat_minutes - (int)lat_minutes) * 1.0E5));
+
+    // ground speed
+    const Vector2f speed = AP::ahrs().groundspeed_vector();
+    float speed_knots = norm(speed.x, speed.y) * M_PER_SEC_TO_KNOTS;
+    snprintf((char*)&pkt.payload[21], 7, "%03u.%02u", (unsigned)speed_knots, unsigned((speed_knots - (int)speed_knots) * 1.0E2));
+
+    // heading
+    //float heading = wrap_360(degrees(atan2f(speed.x, speed.y)));
+    float heading = wrap_360(degrees(speed.angle()));
+    snprintf((char*)&pkt.payload[27], 10, "%03u.%04u", unsigned(heading), unsigned((heading - (int)heading) * 1.0E4));
+
+    // hemisphere
+    uint8_t hemisphere = 0;
+    hemisphere |= (latitude >= 0) ? 0x01 : 0;   // isNorth
+    hemisphere |= (longitude >= 0) ? 0x02 : 0;  // isEast
+    hemisphere |= (AP::gps().status() < AP_GPS::GPS_OK_FIX_2D) ? 0x80 : 0;  // isInvalid
+    pkt.payload[35] = hemisphere;
+
+    // time
+    uint64_t time_usec;
+    if (!AP::rtc().get_utc_usec(time_usec)) {
+        memset(&pkt.payload[36],' ', 10);
+    } else {
+        // not completely accurate, our time includes leap seconds and time_t should be without
+        const time_t time_sec = time_usec / 1000000;
+        struct tm* tm = gmtime(&time_sec);
+
+        // format time string
+        snprintf((char*)&pkt.payload[36], 11, "%02u%02u%06.3f", tm->tm_hour, tm->tm_min, tm->tm_sec + (time_usec % 1000000) * 1.0e-6);
+
+    }
+
+    send_msg(pkt);
+}
+
+const char* AP_ADSB_Sagetech::type_to_str(const uint8_t type)
+{
+    switch (type) {
+    case 0x01: return "Install";
+    case 0x02: return "PreFlight";
+    case 0x03: return "Operate";
+    case 0x04: return "GPS Data";
+    case 0x05: return "DataReq";
+    default:   return "Unknown";
+    }
+}
+
+const char* AP_ADSB_Sagetech::systemStatsBits_to_str(const SystemStateBits systemStateBits)

```

Comment:
This really doesn't do what its name hints it does - this is a bitmask and this won't return you anything useful if two bits are set.

Reply:
It should also be static